connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java

```diff
@@ -0,0 +1,144 @@
+                        resolvedSchema.getPrimaryKeyIndexes().length));
+        int rowKeyIndex = resolvedSchema.getPrimaryKeyIndexes()[0];
+        checkArgument(
+                resolvedSchema
+                        .getColumn(rowKeyIndex)
+                        .get()
+                        .getDataType()
+                        .equals(DataTypes.STRING().notNull()),
+                String.format(
+                        ErrorMessages.ROW_KEY_STRING_TYPE_TEMPLATE,
+                        resolvedSchema.getColumn(rowKeyIndex).get().getDataType()));
+
+        this.connectorOptions = connectorOptions;
+        this.resolvedSchema = resolvedSchema;
+        this.parallelism = connectorOptions.get(BigtableConnectorOptions.SINK_PARALLELISM);
+        this.rowKeyField = resolvedSchema.getColumn(rowKeyIndex).get().getName();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.resolvedSchema, this.connectorOptions);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        BigtableDynamicTableSink other = (BigtableDynamicTableSink) obj;
+        return Objects.equals(resolvedSchema, other.resolvedSchema)
+                && Objects.equals(connectorOptions, other.connectorOptions);
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        return ChangelogMode.insertOnly();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+        DataType physicalSchema = resolvedSchema.toPhysicalRowDataType();
+
+        RowDataToRowMutationSerializer.Builder serializerBuilder =
+                RowDataToRowMutationSerializer.builder()
+                        .withSchema(physicalSchema)
+                        .withRowKeyField(this.rowKeyField);
+
+        if (connectorOptions.get(BigtableConnectorOptions.USE_NESTED_ROWS_MODE)) {
+            serializerBuilder.withNestedRowsMode();
+        } else {
+            serializerBuilder.withColumnFamily(
+                    connectorOptions.get(BigtableConnectorOptions.COLUMN_FAMILY));
+        }
+
+        final BigtableSink<RowData> bigtableSink =
+                BigtableSink.<RowData>builder()
+                        .setProjectId(connectorOptions.get(BigtableConnectorOptions.PROJECT))
+                        .setTable(connectorOptions.get(BigtableConnectorOptions.TABLE))
+                        .setInstanceId(connectorOptions.get(BigtableConnectorOptions.INSTANCE))
+                        .setSerializer(serializerBuilder.build())
+                        .build();
+
+        if (this.parallelism == null) {
+            return SinkV2Provider.of(bigtableSink);
+        }
+        return SinkV2Provider.of(bigtableSink, this.parallelism);
```

Comment:
i dont have a preference on if you explicitly reference this.variable vs just using variable for class variables but we should keep it consistent (ie, on like 137, we should use this.)

Reply:
Thanks!